src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/PhysicalProjectTreeStorage.cs

```diff
@@ -22,34 +23,52 @@ internal class PhysicalProjectTreeStorage : IPhysicalProjectTreeStorage
                                           [Import(ExportContractNames.ProjectTreeProviders.PhysicalViewTree)]Lazy<IProjectTreeProvider> treeProvider,
                                           Lazy<IFileSystem> fileSystem,
                                           ActiveConfiguredProject<IFolderManager> folderManager,
+                                          [Import(ExportContractNames.ProjectItemProviders.SourceFiles)]ActiveConfiguredProject<IProjectItemProvider> sourceItemProvider,
                                           UnconfiguredProject project)
         {
             _treeService = treeService;
             _treeProvider = treeProvider;
+            _sourceItemProvider = sourceItemProvider;
             _fileSystem = fileSystem;
             _folderManager = folderManager;
             _unconfiguredProject = project;
         }
 
-        public Task<IProjectTree> CreateFolderAsync(string path)
+        public async Task<IProjectItem?> CreateFileAsync(string path)
         {
             Requires.NotNullOrEmpty(path, nameof(path));
 
-            if (_treeService.Value.CurrentTree == null)
-                throw new InvalidOperationException("Physical project tree has not yet been published.");
+            return await AddToProjectAsync(path, waitForFileSystemUpdates: true, async fullPath =>
+            {
+                using (_fileSystem.Value.Create(fullPath)) { }
 
-            string fullPath = _unconfiguredProject.MakeRooted(path);
+                await _sourceItemProvider.Value.AddAsync(fullPath);
```

Comment:
Can remote this async state machine, and maybe the outer one, depending on that cast.